### PR TITLE
561380 - API revision: requirements resolution

### PR DESCRIPTION
**Pull request by eparovyshnaya**
Bug 561380 - API revision: requirements resolution

 - Licensing requirement and resolution interfaces are revised. Existing
ones are deprecated in favor of new yet `internal` variants
 - Equinox-based resolution are under revision. The first one (
`manifest`-sourced) resolver is migrating to new api and intermediate
services
  - related static utils and public static literals are deprecated and
substituted with dynamic constructions

Signed-off-by: elena.parovyshnaya <elena.parovyshnaya@gmail.com>

### CODE REVIEW

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/BaseFeature.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+@SuppressWarnings("restriction")
+public final class BaseFeature implements Feature {
+
+	private final String id;
+	private final String version;
+	private final String name;
+	private final String provider;
+
+	public BaseFeature(String id, String version, String name, String provider) {
+		Objects.requireNonNull(id);
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/BaseFeature.java]: user readable messages

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/BaseRequirement.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+@SuppressWarnings("restriction")
+public final class BaseRequirement implements Requirement {
+
+	private final Feature feature;
+	private final RestrictionLevel restriction;
+	private final Object source;
+
+	public BaseRequirement(Feature feature, RestrictionLevel restriction, Object source) {
+		Objects.requireNonNull(feature);
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/BaseRequirement.java]: user readable messages

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		this.source = source;
+	}
+
+	@Override
+	public Requirement get() {
+		Feature feature = new BaseFeature( //
+				Long.toHexString(System.currentTimeMillis()), //
+				"0.0.0", //$NON-NLS-1$
+				description, //
+				"Passage License Management"); //$NON-NLS-1$
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java]: i18n

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	private final String description;
+	private final Object source;
+
+	/**
+	 * @param description to be used as {@code feature name}, descriptive
+	 * @param source      physical source of an configuration error, to be used as a
+	 *                    {@code requirement definition source}
+	 */
+	public UnsatisfiableRequirement(String description, Object source) {
+		this.description = description;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java]: null checks with i18n

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
+package org.eclipse.passage.lic.internal.base.requirements;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/requirements/UnsatisfiableRequirement.java]: header

File: bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/restrictions/DefaultRestrictionLevel.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.eclipse.passage.lic.internal.base.restrictions;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.base/src/org/eclipse/passage/lic/internal/base/restrictions/DefaultRestrictionLevel.java]: header

File: bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/RequirementsSupplierRegistry.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import org.eclipse.passage.lic.internal.api.registry.StringServiceId;
+import org.eclipse.passage.lic.internal.api.requirements.ResolvedRequirements;
+
+@SuppressWarnings("restriction")
+public final class RequirementsSupplierRegistry implements Supplier<Registry<StringServiceId, ResolvedRequirements>> {
+
+	@Override
+	public Registry<StringServiceId, ResolvedRequirements> get() {
+
+		return null;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/RequirementsSupplierRegistry.java]: hmm. what this strange literal stands for?

File: bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/LicensingFeaturesFromBundle.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package org.eclipse.passage.lic.internal.equinox.requirements;
```
Comment (ruspl-afed) [inline bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/requirements/LicensingFeaturesFromBundle.java]: header

File: bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/internal/api/requirements/ResolvedRequirements.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+ * <p>
+ * At the program runtime a {@code RequirementResolver} reads particular type of
+ * physical sources and produce set of {@link Requirement}s that are defined
+ * there.
+ * </p>
+ *
+ * @see Requirement
+ * @see org.eclipse.passage.lic.api
+ */
+public interface ResolvedRequirements extends Service<StringServiceId> {
```
[the remark this answers is not in the document]
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.api/src/org/eclipse/passage/lic/internal/api/requirements/ResolvedRequirements.java]: No - it inherits some methods from `Service` interface is extends. Thus it cannot be declared as FunctionalInterface.

File: bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/RequirementsSupplierRegistry.java (hunk 7 above)
Comment (eparovyshnaya) [reply bundles/org.eclipse.passage.lic.equinox/src/org/eclipse/passage/lic/internal/equinox/RequirementsSupplierRegistry.java]: It's autogenerated!  Sorry! Wiped!

**Pull request merged**